### PR TITLE
custom_image_host: Update custom images information when requested

### LINKED ISSUE #569: Transient "Downloaded image hash does not match" error
People have stumbled upon:

```
$ multipass launch snapcraft:core
Launching a VM.
launch failed: Downloaded image hash does not match
```

The likely issue is that somehow we get out of sync with the [`SHA256SUMS`](http://cloud-images.ubuntu.com/minimal/releases/xenial/release/SHA256SUMS) file and trying to compare a previous hash to a new image.

### PR DESCRIPTION
**Pull request by townsend2010**
Fixes #569

### CODE REVIEW

**Comment by townsend2010** [comment]
@ricab,

Well,`VMImageHost` is an interface, so putting common code here will blow up that concept.  I understand what you mean on there is commonality here (and many other places), but we designed many aspects of Multipass to use interfaces and then put the implementation in the derived class.

I'm open to discussing changing this and perhaps we should do that in Malta.  But changing will cause quite a bit of refactoring and I'm not sure if we have the resources for that right now.

But for now, I'm not akin to changing the code design for this fix.

**Comment by ricab** [comment]
@townsend2010 

If we decided (e.g. in Malta) to dedicate some quality time to existing code, I would be happy. But irrespective of that, I think we should strive not to introduce *new* repetition.

I think a sensible approach is to do the refactoring that is strictly required to avoid any new piece of code increasing duplication. Of course, seeing how best to achieve that and considering effort on a case-by-case basis.

As discussed in IRC, I think the intermediate `CommonVMImageHost` that @Saviq suggested is the way to go on this occasion. I will pick up this PR and try to do that. The new class will only get the updating code for now. Other common code that could go in there is deferred to separate PRs.

File: src/daemon/custom_image_host.h [hunk 1]
```diff
@@ -52,9 +53,13 @@ class CustomVMImageHost final : public VMImageHost
     std::vector<std::string> supported_remotes() override;
 
 private:
+    void update_manifest();
     CustomManifest* manifest_from(const std::string& remote_name);
+    std::chrono::seconds manifest_time_to_live;
```
Comment (ricab) [inline src/daemon/custom_image_host.h]: @townsend2010 Just noticed this is being default constructed (TTL=0). This was just an oversight, correct?

Comment (townsend2010) [reply src/daemon/custom_image_host.h]: @ricab,

Yes, you are correct, that is wrong.  In the `UbuntuVMImageHost` ctor, that value is passed in and assigned and the same should be done for the `CustomVMImageHost` or whatever common class you are implementing.

**Comment by ricab** [comment]
OK I just pushed something that is kind of half-way to where I think we should be. I want to request feedback on whether or not I should move on to the next part:

1. Merge `CustomManifest` with `SimpleStreamsManifest`
2. Use the same type to map remotes (i.e. "snapcraft"/"daily"/"release") to manifests (ATM, one specialization uses `std::unordered_map`, the other a vector of pairs)
3. Move the manifest structure to the common base, so that we can do away with the [new *virtuals*](https://github.com/CanonicalLtd/multipass/blob/a59c899bdec8d1988c423b6ef7978d3cceec77fa/src/daemon/common_image_host.h#L38). Populating the manifests would still be left to specializations.

This would simplify things a lot, but require a bit more effort. WDYT?

**Comment by townsend2010** [comment]
@ricab,

Thanks for taking this on.  The new common code is better:)

Regarding your questions, I think we should wait on doing any of that.  But I do think we should tackle that at some point and I'm thinking that is when we put in the actual support for custom image repositories the user can set up.

**Pull request merged**